packages/paste-core/components/combobox/src/Combobox.tsx

```diff
@@ -77,13 +87,61 @@ const Combobox = React.forwardRef<HTMLInputElement, ComboboxProps>(
     } = useComboboxPrimitive({
       initialSelectedItem,
       items,
-      itemToString: itemToString || defaultItemToString,
       onHighlightedIndexChange,
       onInputValueChange,
       onIsOpenChange,
       onSelectedItemChange,
+      ...(itemToString && {itemToString}),
+      ...(initialIsOpen && {initialIsOpen}),
+      ...(inputValue && {inputValue}),
+      ...(selectedItem && {selectedItem}),
     });
     const helpTextId = useUID();
+    const optionUID = useUIDSeed();
+    const groupUID = useUIDSeed();
+
+    const renderItem = (item: any, index: number, inGroup?: boolean): React.ReactNode => {
+      return (
+        <ComboboxListboxOption
+          {...getItemProps({item, index})}
+          highlighted={highlightedIndex === index}
+          key={optionUID(index)}
+          variant={inGroup ? 'groupOption' : 'default'}
+        >
+          {optionTemplate ? optionTemplate(item) : item}
+        </ComboboxListboxOption>
+      );
+    };
+
+    const renderItems = (): React.ReactNode[] =>
+      items.map((item, index) => {
+        return renderItem(item, index);
+      });
+
+    const renderGroupedItems = (): React.ReactNode[] | null => {
```

Comment:
Would be great to move this and the previous functions out of the component. It means these wont reinitialze every re-render, but also it will force us to pass arguments to these functions, making them pure and generally more testable.